### PR TITLE
@UpdateDateColumn support for Upsert

### PR DESCRIPTION
**Pull request by zevstravitz**
### Description of change
tldr; This PR fixes upsert not automatically updating an UpdateDate column by adding any UpdateDate column to the list of overwrite columns for the upsert method.

If you provide an UpdateColumn column name to an orUpdate method's overwrite columns, you don't actually have to provide the new value because typeorm infers it. But with upsert, we update a list of columns implicitly provided from our list of entities, which should not have to include an UpdateDate column value. This PR resolves this problem by _automatically_ adding any UpdateDate column to the list of overwrite columns for the upsert method.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [ ] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

fix: Adds support for @UpdateDateColumn support for Upsert method

Closes: #9015

### CODE REVIEW

**Comment by pleerock** [comment]
Looks like tests are failing.

**Comment by Thore1954** [comment]
Any updates on this?

**Comment by Thore1954** [comment]
@pleerock I manually tested this fork and it's working. I think the tests are failing because they don't expect the update date column to be updated!
My current work around is:
```ts
await repository.upsert(
    [
        { externalId: "abc123", firstName: "Rizzrak", updatedAt: () => `now()` },
        { externalId: "bca321", firstName: "Karzzir", updatedAt: () => `now()` },
    ],
    ["externalId"],
)
```
But it only works on postgres so would love to get rid of it as soon as possible.

**Comment by Thore1954** [comment]
Actually, there's currently no assertion on the update date column in tests:
https://github.com/typeorm/typeorm/blob/85fa9c6e7df1a0e3ba8c95d64d9b2a6ab464e0e0/test/functional/repository/basic-methods/repository-basic-methods.ts#L500-L505
Why are the tests failing? @zevstravitz @pleerock

**Comment by Thore1954** [comment]
I found out the problem is the `skipUpdateIfNoValuesChanged` option is not respected by this change.
One solution is to exclude update date column from the condition:
https://github.com/typeorm/typeorm/blob/0d72317b3db3f93ce9d4c2f6f027c7486bbed08b/src/query-builder/InsertQueryBuilder.ts#L542-L551
Should be changed to:
```ts
                    query += overwrite
                        .filter(column => !(this.expressionMap.mainAlias!.hasMetadata && this.expressionMap.mainAlias!.metadata.findColumnWithDatabaseName(column)!.isUpdateDate))
                        .map(
                            (column) =>
                                `${tableName}.${this.escape(
                                    column,
                                )} IS DISTINCT FROM EXCLUDED.${this.escape(
                                    column,
                                )}`,
                        )
                        .join(" OR ")
```

File: src/entity-manager/EntityManager.ts [hunk 1]
```diff
@@ -704,10 +704,11 @@ export class EntityManager {
         const overwriteColumns = metadata.columns.filter(
             (col) =>
                 !conflictColumns.includes(col) &&
-                entities.some(
+                (entities.some(
                     (entity) =>
                         typeof col.getEntityValue(entity) !== "undefined",
-                ),
+                ) ||
+                    col.isUpdateDate),
```
Comment (Thore1954) [inline src/entity-manager/EntityManager.ts:711]: When `skipUpdateIfNoValuesChanged` option is set, it shouldn't add update date column otherwise it will update even when nothing has changed.

**Review by Thore1954 — CHANGES_REQUESTED** [review]

**Comment by Thore1954** [comment]
@zevstravitz thanks for the PR! Unfortunately, this wouldn't work because when `skipUpdateIfNoValuesChanged` option is set it won't skip update when nothing has changed.

**Comment by AlexMesser** [comment]
Closing due to inactivity and failing tests. Feel free to submit a new PR.
